AABB_tree/include/CGAL/AABB_tree/internal/triangle_datum_covering.h

```diff
@@ -0,0 +1,493 @@
+    {
+      // check a datum only once
+      auto is_insert_successful = visited_data.insert(primitive.id().second/*unique input face ID*/);
+      if(!is_insert_successful.second)
+        return;
+
+      return Base::intersection(query, primitive);
+    }
+  };
+
+  template <typename Query, typename CountingIterator>
+  class Listing_intersection_traits
+    : public CGAL::internal::AABB_tree::Listing_intersection_traits<AABBTraits, Query, CountingIterator>
+  {
+    using Base = CGAL::internal::AABB_tree::Listing_intersection_traits<AABBTraits, Query, CountingIterator>;
+    using Primitive = typename AABBTraits::Primitive;
+
+    std::unordered_set<std::size_t> visited_data;
+
+  public:
+    template <typename OutputIterator>
+    Listing_intersection_traits(OutputIterator out, const AABBTraits& traits) : Base(out, traits) { }
+
+    void intersection(const Query& query, const Primitive& primitive)
+    {
+      // check a datum only once
+      auto is_insert_successful = visited_data.insert(primitive.id().second/*unique input face ID*/);
+      if(!is_insert_successful.second)
+        return;
+
+      return Base::intersection(query, primitive);
+    }
+  };
+};
+
+// Dissociated from the class `AABB_covered_triangle_tree` for clarity
+template <typename Kernel, typename Point>
+struct AABB_covered_triangle_tree_traits
+{
+  using Triangle_3 = typename Kernel::Triangle_3;
+
+  // Below is a lot of trouble to cover a single datum with multiple primitives using smaller bboxes
+  using ID = std::pair<std::size_t /*primitive ID*/, std::size_t /*input face ID*/>;
+  using IDPM = CGAL::First_of_pair_property_map<ID>;
+
+  // Primitive ID --> box vector pos --> Bounding Box
+  using BPMB = internal::Vector_property_map<CGAL::Bbox_3>;
+  using BPM = CGAL::Property_map_binder<IDPM, BPMB>;
+
+  // Primitive ID --> point vector pos --> Reference Point
+  using RPPMB = internal::Vector_property_map<Point>;
+  using RPPM = CGAL::Property_map_binder<IDPM, RPPMB>;
+
+  // Primitive ID --> Datum pos vector pos --> Datum pos --> Datum
+  // The vector of data has size nf, but the vector of datum pos has size tree.size()
+  using DPPMB = internal::Vector_property_map<std::size_t>; // pos --> Datum pos
+  using DPPM = CGAL::Property_map_binder<IDPM, DPPMB>; // PID --> Datum pos
+  using DPMB = internal::Vector_property_map<Triangle_3>; // Datum pos --> Datum
+  using DPM = CGAL::Property_map_binder<DPPM, DPMB>; // PID --> Datum
+
+  using Primitive = CGAL::AABB_primitive<ID, DPM, RPPM,
+                                         CGAL::Tag_true /*external pmaps*/,
+                                         CGAL::Tag_false /*no caching*/>;
+
+  using AABB_geom_traits = Kernel;
+  using AABB_traits = CGAL::AABB_traits<AABB_geom_traits, Primitive, BPM>;
+  using AABB_tree = CGAL::AABB_tree<AABB_traits>;
+};
+
+template <typename Kernel,
+          typename Point = typename Kernel::Point_3>
+struct AABB_covered_triangle_tree
```

Comment:
I don't like inheritance. In principle all could be done in a free function with the extra data being stored as shared data in the traits. This is just a suggestion, not an implementation request.

Reply:
Also, we can add a new `insert()` function to the AABB-tree than would not blindly take the input range but process it before doing the real insert. This could be use to augment that range or to reduce it (think about removing degenerate faces for example or filtering out some elements)

Reply:
This discussion is not marked as resolved. @MaelRL @sloriot is that still something you want to discuss?